<a href="https://colab.research.google.com/github/cu-applied-math/appm-4600-numerics/blob/main/Demos/Ch1_SymbolicTaylorSeries.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

Manipulate Taylor series using the **SymPy** package in Python, following the tutorial at [scipy-lectures.org/packages/sympy.html](https://scipy-lectures.org/packages/sympy.html).  SymPy is used in [SageMath](https://www.sagemath.org/).

You can do this in Maple, which is a bit old-school, or Mathematica (which we have a license for at CU).  Wolfram Alpha is another choice (by the company that makes Mathematica) though the free version does have some limitations.

The nice thing about Python is that it is free and community supported, so if you take the time to use it, it won't go away. If you learn Mathematica now, you may not be able to use it in your job, since it's an expensive license.

In [3]:
# Load the package
import numpy as np
import sympy as sym
from sympy import init_printing
init_printing()  # This will make output look very nice!

import math  # has things like math.pi (which is the numerical value of pi)

In [4]:
# First, all symbolic variables must be declared. This is in contrast to Mathematica
x = sym.Symbol('x')

# Now, let's ask for a Taylor series expansion
sym.series( sym.cos(x), x )

     2    4        
    x    x     ⎛ 6⎞
1 - ── + ── + O⎝x ⎠
    2    24        

In [ ]:
# We can see the Docstring for the function to see all the options
?sym.series

Before going further, let's play with cos(pi/4)

In [5]:
print( math.cos( math.pi/4 ) )
# or, ask to print out 20 decimal places
print( '%.20f' % math.cos( math.pi/4 ) )  # this is old-school: see https://docs.python.org/3/tutorial/inputoutput.html#old-string-formatting
print( math.cos( sym.pi/4) )
print( sym.cos( math.pi/4) )
print( sym.cos( sym.pi/4) )
sym.cos( sym.pi/4)  # same as above, but use fancy formatting

0.7071067811865476
0.70710678118654757274
0.7071067811865476
0.707106781186548
sqrt(2)/2


√2
──
2 

and even simpler, play around with $\pi$.  The default Python math library will give us as many digits as we request... but are they accurate digits?

In [6]:
print(f'{math.pi:.30f}') # 30 digits of pi... ?!
print(f'{sym.pi:.30f}') # 30 digits of pi... ?!
print(f'{3.141592653589793238462643383279:.30}') # from https://www.piday.org/million/
print('3.141592653589793238462643383279')
# What is going on!?  Even if we have an accurate approximation, it's hard for numpy to display it!

3.141592653589793115997963468544
3.141592653589793238462643383279
3.14159265358979311599796346854
3.141592653589793238462643383279


In [9]:
# Numpy printing is a little smarter: if we request 40 digits of precision
# but it knows that the number doesn't have that many *accurate* digits,
# then it will refuse to print them
with np.printoptions(precision=40):
    print(np.pi)
    print(math.pi)
    print(sym.pi)
    print(3.141592653589793238462643383279)
    print('3.141592653589793238462643383279')

3.141592653589793
3.141592653589793
pi
3.141592653589793
3.141592653589793238462643383279


Now let's explore the series. In particular, let's try to evaluate cos( .75 ), using only addition and multiplication, and a few "memorized" values of sine/cosine (let's assume that we memorize the fact that $\cos( \pi/4 ) = \sqrt{2}/2$ and that $\sin( \pi/4 ) = \sqrt{2}/2$.

Since .75 is not that far from $\pi/4 \approx 0.785$, let's do a Taylor expansion of $\cos$ around $\pi/4$.

In [ ]:
Taylor = sym.series( sym.cos(x), x, x0=sym.pi/4, n=5)
Taylor

                            2             3             4                     
        ⎛    π⎞      ⎛    π⎞       ⎛    π⎞       ⎛    π⎞                      
     √2⋅⎜x - ─⎟   √2⋅⎜x - ─⎟    √2⋅⎜x - ─⎟    √2⋅⎜x - ─⎟     ⎛       5       ⎞
√2      ⎝    4⎠      ⎝    4⎠       ⎝    4⎠       ⎝    4⎠     ⎜⎛    π⎞       π⎟
── - ────────── - ─────────── + ─────────── + ─────────── + O⎜⎜x - ─⎟ ; x → ─⎟
2        2             4            12            48         ⎝⎝    4⎠       4⎠

In [ ]:
Taylor.subs(x,.75) # Not what we want

O(1)

In [ ]:
Taylor.evalf( subs={x: .75})

                            2             3             4                     
        ⎛    π⎞      ⎛    π⎞       ⎛    π⎞       ⎛    π⎞                      
     √2⋅⎜x - ─⎟   √2⋅⎜x - ─⎟    √2⋅⎜x - ─⎟    √2⋅⎜x - ─⎟     ⎛       5       ⎞
√2      ⎝    4⎠      ⎝    4⎠       ⎝    4⎠       ⎝    4⎠     ⎜⎛    π⎞       π⎟
── - ────────── - ─────────── + ─────────── + ─────────── + O⎜⎜x - ─⎟ ; x → ─⎟
2        2             4            12            48         ⎝⎝    4⎠       4⎠

In [ ]:
# Numerically evaluating this was a problem, because of the O( x^5 ) term. Let's remove it.
# See https://docs.sympy.org/latest/tutorial/calculus.html
Taylor.removeO()

          4             3             2                  
   ⎛    π⎞       ⎛    π⎞       ⎛    π⎞       ⎛    π⎞     
√2⋅⎜x - ─⎟    √2⋅⎜x - ─⎟    √2⋅⎜x - ─⎟    √2⋅⎜x - ─⎟     
   ⎝    4⎠       ⎝    4⎠       ⎝    4⎠       ⎝    4⎠   √2
─────────── + ─────────── - ─────────── - ────────── + ──
    48            12             4            2        2 

In [ ]:
guess = Taylor.removeO().subs(x,.75).evalf()
print("Our guess is: \t%.15f" % guess)   # .15f means give us a fixed point representation, with 15 digits to the right of the .
print("True value is: \t%.15f" % math.cos( .75) )
print(f"Discrepency is: \t{abs( guess - math.cos(.75) ):.3e}")  # .3e means give us scientifc notation with 3 digits to the right of the .

Our guess is: 	0.731688868548266
True value is: 	0.731688868873821
Discrepency is: 	3.256e-10


Our *absolute* error is about $3\times 10^{-10}$. Is this what we'd expect? Let's use math to be able to **guarantee** a small error (without having known the true answer first)

Let's look at the remainder term in the Taylor series

In [ ]:
x0=sym.Symbol('x0')
sym.series( sym.cos(x), x, x0=x0, n=5)

                                     2                   3                   4 ↪
                             (x - x₀) ⋅cos(x₀)   (x - x₀) ⋅sin(x₀)   (x - x₀)  ↪
cos(x₀) - (x - x₀)⋅sin(x₀) - ───────────────── + ───────────────── + ───────── ↪
                                     2                   6                  24 ↪

↪                                
↪ ⋅cos(x₀)    ⎛        5        ⎞
↪ ──────── + O⎝(x - x₀) ; x → x₀⎠
↪                                

In [ ]:
xi = sym.Symbol('xi')
Remainder = -(x-x0)**5*sym.sin(xi)/120
Remainder

         5        
-(x - x₀) ⋅sin(ξ) 
──────────────────
       120        

By Taylor's remainder theorem, we know $\xi \in [x,x_0]$ (or $\xi \in [x_0,x]$ if $x_0 < x$) such that the above remainder is the error in the Taylor series. We don't know exactly what $\xi$ is, but that's OK, because for any $\xi$, we can just bound $|\sin(\xi)| \le 1$.  So using this,...

In [ ]:
RemainderBound = -(x-x0)**5/120
RemainderBound

         5 
-(x - x₀)  
───────────
    120    

In [ ]:
# See https://docs.sympy.org/latest/tutorial/basic_operations.html
RemainderBound.subs( [ (x,.75), (x0,math.pi/4) ])

4.63150782430020e-10

To summarize, our Taylor series approximation had error $3.25 \times 10^{-10}$.  We were able to use math to give an *a priori* bound that the error was no more than $4.63 \times 10^{-10}$ (though in that bound, we disregarded potentially cancellation errors when summing the series)